File: ui/src/utils/socialSecurity.js
/**
 * Social Security Benefit Calculator (Frontend)
 * 
 * Calculates Social Security benefits based on:
 * - Full Retirement Age (FRA) based on birth year
 * - Primary Insurance Amount (PIA)
 * - Retirement date (early, on-time, or delayed)
 * - COLA adjustments
 */

export function calculateFRA(birthYear) {
  /**
   * Calculate Full Retirement Age (FRA) based on birth year.
   * Returns an object with years and months.
   */
  if (birthYear >= 1960) {
    return { years: 67, months: 0 };
  } else if (birthYear === 1959) {
    return { years: 66, months: 10 };
  } else if (birthYear === 1958) {
    return { years: 66, months: 8 };
  } else if (birthYear === 1957) {
    return { years: 66, months: 6 };
  } else if (birthYear === 1956) {
    return { years: 66, months: 4 };
  } else if (birthYear === 1955) {
    return { years: 66, months: 2 };
  } else if (birthYear >= 1943 && birthYear <= 1954) {
    return { years: 66, months: 0 };
  } else {
    // For years before 1943, FRA is 65
    return { years: 65, months: 0 };
  }
}

export function calculateFRADate(birthDate) {
  /**
   * Calculate the Full Retirement Age date from birth date.
   * Returns date string in YYYY-MM-DD format or null if invalid.
   */
  if (!birthDate) return null;
  
  try {
    const birth = new Date(birthDate + 'T00:00:00'); // Add time to avoid timezone issues
    const birthYear = birth.getFullYear();
    const fra = calculateFRA(birthYear);
    
    // Calculate FRA date
    const fraDate = new Date(birth);
    fraDate.setFullYear(birth.getFullYear() + fra.years);
    fraDate.setMonth(birth.getMonth() + fra.months);
    
    // Format as YYYY-MM-DD
    const year = fraDate.getFullYear();
    const month = String(fraDate.getMonth() + 1).padStart(2, '0');
    const day = String(fraDate.getDate()).padStart(2, '0');
    
    return `${year}-${month}-${day}`;
  } catch (e) {
    return null;
  }
}

export function formatFRADisplay(birthDate) {
  /**
   * Format FRA for display (e.g., "66 years, 2 months" or "67 years")
   */
  if (!birthDate) return '';
  
  try {
    const birth = new Date(birthDate + 'T00:00:00');
    const birthYear = birth.getFullYear();
    const fra = calculateFRA(birthYear);
    
    if (fra.months === 0) {
      return `${fra.years} years`;
    } else {
      return `${fra.years} years, ${fra.months} months`;
    }
  } catch (e) {
    return '';
  }
}

export function calculateMonthlyBenefit(pia, retirementDate, fraDate, birthDate) {
  /**
   * Calculate Social Security monthly benefit based on PIA, retirement date, and FRA.
   * This is a simplified version - the backend will do the full calculation.
   */
  if (!pia || pia <= 0 || !retirementDate || !fraDate) {
    return pia || 0;
  }
  
  try {
    const retDate = new Date(retirementDate + 'T00:00:00');
    const fra = new Date(fraDate + 'T00:00:00');
    
    // Calculate months difference
    const monthsDiff = (retDate.getFullYear() - fra.getFullYear()) * 12 + 
                       (retDate.getMonth() - fra.getMonth());
    
    if (monthsDiff < 0) {
      // Early retirement
      const monthsEarly = Math.abs(monthsDiff);
      
      let reduction = 0;
      if (monthsEarly <= 36) {
        reduction = monthsEarly * (5/9) * 0.01;
      } else {
        reduction = 36 * (5/9) * 0.01 + (monthsEarly - 36) * (5/12) * 0.01;
      }
      
      return Math.round(pia * (1 - reduction) * 100) / 100;
    } else if (monthsDiff > 0) {
      // Delayed retirement
      // Calculate age at retirement
      let ageAtRetirement = retDate.getFullYear() - fra.getFullYear() + (calculateFRA(new Date(birthDate + 'T00:00:00').getFullYear()).years);
      
      if (!birthDate) {
        ageAtRetirement = 70; // Default
      } else {
        const birth = new Date(birthDate + 'T00:00:00');
        ageAtRetirement = retDate.getFullYear() - birth.getFullYear() - 
          ((retDate.getMonth() < birth.getMonth()) || 
           (retDate.getMonth() === birth.getMonth() && retDate.getDate() < birth.getDate()) ? 1 : 0);
      }
      
      // Delay credits: 2/3 of 1% per month up to age 70
      const monthsDelay = ageAtRetirement >= 70 ? 
        Math.min(monthsDiff, 36) : // Approximate max 36 months
        monthsDiff;
      
      const increase = monthsDelay * (2/3) * 0.01;
      return Math.round(pia * (1 + increase) * 100) / 100;
    } else {
      // Exactly at FRA
      return pia;
    }
  } catch (e) {
    return pia || 0;
  }
}

export function getMinRetirementDate(birthDate) {
  /**
   * Get minimum retirement date (62nd birthday) for validation.
   * Returns date string in YYYY-MM-DD format or null.
   */
  if (!birthDate) return null;
  
  try {
    const birth = new Date(birthDate + 'T00:00:00');
    const minDate = new Date(birth);
    minDate.setFullYear(birth.getFullYear() + 62);
    
    const year = minDate.getFullYear();
    const month = String(minDate.getMonth() + 1).padStart(2, '0');
    const day = String(minDate.getDate()).padStart(2, '0');
    
    return `${year}-${month}-${day}`;
  } catch (e) {
    return null;
  }
}

File: api/utils/social_security.py
"""
Social Security Benefit Calculator

Calculates Social Security benefits based on:
- Full Retirement Age (FRA) based on birth year
- Primary Insurance Amount (PIA)
- Retirement date (early, on-time, or delayed)
- COLA adjustments

Based on Social Security Administration rules.
"""

from datetime import date, datetime
from typing import Tuple, Optional


def calculate_fra(birth_year: int) -> Tuple[int, int]:
    """
    Calculate Full Retirement Age (FRA) based on birth year.
    
    Args:
        birth_year: Year of birth
    
    Returns:
        Tuple of (years, months) representing FRA
    """
    if birth_year >= 1960:
        return (67, 0)
    elif birth_year == 1959:
        return (66, 10)
    elif birth_year == 1958:
        return (66, 8)
    elif birth_year == 1957:
        return (66, 6)
    elif birth_year == 1956:
        return (66, 4)
    elif birth_year == 1955:
        return (66, 2)
    elif 1943 <= birth_year <= 1954:
        return (66, 0)
    else:
        # For years before 1943, FRA is 65
        return (65, 0)


def fra_to_date(birth_date: str) -> Optional[date]:
    """
    Calculate the Full Retirement Age date from birth date.
    
    Args:
        birth_date: Birth date in YYYY-MM-DD format
    
    Returns:
        FRA date as date object, or None if invalid
    """
    try:
        birth = datetime.strptime(birth_date, "%Y-%m-%d").date()
        birth_year = birth.year
        fra_years, fra_months = calculate_fra(birth_year)
        
        # Calculate FRA date
        fra_date = date(
            birth.year + fra_years,
            birth.month + fra_months,
            birth.day
        )
        # Handle month overflow (e.g., if month > 12)
        if fra_date.month > 12:
            fra_date = date(
                fra_date.year + 1,
                fra_date.month - 12,
                fra_date.day
            )
        
        return fra_date
    except (ValueError, AttributeError):
        return None


def calculate_monthly_benefit(
    pia: float,
    retirement_date: str,
    fra_date: Optional[date],
    birth_date: Optional[str] = None
) -> float:
    """
    Calculate Social Security monthly benefit based on PIA, retirement date, and FRA.
    
    Args:
        pia: Primary Insurance Amount (full retirement benefit)
        retirement_date: Retirement date in YYYY-MM-DD format
        fra_date: Full Retirement Age date (can be calculated if None and birth_date provided)
        birth_date: Birth date in YYYY-MM-DD format (used if fra_date is None)
    
    Returns:
        Monthly benefit amount
    """
    if pia <= 0:
        return 0.0
    
    try:
        ret_date = datetime.strptime(retirement_date, "%Y-%m-%d").date()
    except (ValueError, AttributeError):
        return pia  # Return PIA if invalid date
    
    # Calculate FRA date if not provided
    if fra_date is None and birth_date:
        fra_date = fra_to_date(birth_date)
    
    if fra_date is None:
        return pia  # Return PIA if we can't calculate FRA
    
    # Calculate months between retirement date and FRA
    months_diff = (ret_date.year - fra_date.year) * 12 + (ret_date.month - fra_date.month)
    
    # Adjust benefit based on retirement date relative to FRA
    if months_diff < 0:
        # Early retirement (before FRA)
        months_early = abs(months_diff)
        
        # Reduction: 5/9 of 1% per month up to 36 months
        reduction = 0.0
        if months_early <= 36:
            reduction = months_early * (5/9) * 0.01
        else:
            # 5/9 of 1% for first 36 months
            reduction = 36 * (5/9) * 0.01
            # 5/12 of 1% for each additional month
            reduction += (months_early - 36) * (5/12) * 0.01
        
        benefit = pia * (1 - reduction)
        
    elif months_diff > 0:
        # Delayed retirement (after FRA)
        # Calculate age at retirement
        if birth_date:
            try:
                birth = datetime.strptime(birth_date, "%Y-%m-%d").date()
                age_at_retirement = (ret_date.year - birth.year) - ((ret_date.month, ret_date.day) < (birth.month, birth.day))
            except (ValueError, AttributeError):
                age_at_retirement = 70  # Default to 70
        else:
            # Estimate age at retirement (assume they were born at FRA - 66 or 67)
            age_at_retirement = fra_date.year - (fra_date.year - 67) + (ret_date.year - fra_date.year)
        
        # Delay credits: 2/3 of 1% per month for every month past FRA until age 70
        # After age 70, no additional credits
        if age_at_retirement >= 70:
            months_delay = min(months_diff, (70 - (fra_date.year - birth.year)) * 12) if birth_date else min(months_diff, 36)
        else:
            months_delay = months_diff
        
        increase = months_delay * (2/3) * 0.01
        benefit = pia * (1 + increase)
        
    else:
        # Exactly at FRA
        benefit = pia
    
    return round(benefit, 2)


def calculate_spousal_benefit(
    person1_pia: float,
    person2_retirement_date: str,
    person2_fra_date: Optional[date] = None,
    person2_birth_date: Optional[str] = None
) -> float:
    """
    Calculate Person 2's spousal Social Security benefit (based on Person 1's PIA).
    
    Spousal benefit is 32.5-50% of Person 1's PIA, depending on when Person 2 claims
    relative to Person 2's own FRA (not Person 1's FRA).
    
    Base amount: 50% of Person 1's PIA
    - If claiming before Person 2's FRA:
      * First 36 months: reduced by 25/36 of 1% per month
      * After 36 months: reduced by 5/12 of 1% per month
      * Minimum at age 62: 32.5% of Person 1's PIA
    - If claiming at or after Person 2's FRA: 50% of Person 1's PIA
    
    Args:
        person1_pia: Person 1's PIA (Primary Insurance Amount)
        person2_retirement_date: Person 2's retirement date in YYYY-MM-DD format
        person2_fra_date: Person 2's FRA date (optional, calculated if not provided)
        person2_birth_date: Person 2's birth date in YYYY-MM-DD format (used if person2_fra_date is None)
    
    Returns:
        Person 2's monthly spousal benefit amount
    """
    # Calculate Person 2's FRA date if needed
    if person2_fra_date is None and person2_birth_date:
        person2_fra_date = fra_to_date(person2_birth_date)
    
    if person2_fra_date is None:
        # Can't calculate spousal benefit without Person 2's FRA, return 0
        return 0.0
    
    try:
        person2_ret_date = datetime.strptime(person2_retirement_date, "%Y-%m-%d").date()
    except (ValueError, AttributeError):
        return 0.0
    
    # Calculate months between Person 2's retirement and Person 2's own FRA
    months_diff = (person2_ret_date.year - person2_fra_date.year) * 12 + (person2_ret_date.month - person2_fra_date.month)
    
    # Spousal benefit percentage depends on when Person 2 claims relative to Person 2's own FRA
    # Base amount is 50% of Person 1's PIA
    if months_diff < 0:
        # Person 2 claims before their own FRA
        months_early = abs(months_diff)
        
        # Reduction calculation:
        # First 36 months: 25/36 of 1% per month
        # After 36 months: 5/12 of 1% per month
        if months_early <= 36:
            reduction = months_early * (25/36) * 0.01
        else:
            # First 36 months
            reduction = 36 * (25/36) * 0.01
            # Additional months after 36
            reduction += (months_early - 36) * (5/12) * 0.01
        
        # Maximum reduction is to 32.5% (50% - 17.5%)
        reduction = min(reduction, 0.175)
        spousal_percentage = 0.50 - reduction
    else:
        # Person 2 claims at or after their own FRA
        spousal_percentage = 0.50
    
    spousal_benefit = person1_pia * spousal_percentage
    
    return round(spousal_benefit, 2)
